File: src/main/java/com/bytebreakstudios/tictactoe/action/PlayAction.java
package com.bytebreakstudios.tictactoe.action;

import com.bytebreakstudios.tictactoe.State;
import com.bytebreakstudios.board.Action;
import com.bytebreakstudios.board.utils.GameBoardException;

public class PlayAction extends Action<State> {
    private static final String[] validKeys = new String[]{"q", "w", "e", "a", "s", "d", "z", "x", "c"};
    private String key;
    private int player;
    public PlayAction(String key, int player){
        if (key == null) throw new GameBoardException("Key cannot be null");
        if (player == 0) throw new GameBoardException("Player must either be < 0 or > 0 but not == 0");
        this.key = key.toLowerCase();
        this.player = Math.min(1, Math.max(-1, player));

        boolean found = false;
        for (String validKey : validKeys) if(validKey.equals(this.key)) found = true;
        if (!found) throw new GameBoardException("Invalid key: " + key);
    }

    private State doIt(String key, int player, State state){
        switch(key){
            case "q":
                state.q = player;
                break;
            case "w":
                state.w = player;
                break;
            case "e":
                state.e = player;
                break;
            case "a":
                state.a = player;
                break;
            case "s":
                state.s = player;
                break;
            case "d":
                state.d = player;
                break;
            case "z":
                state.z = player;
                break;
            case "x":
                state.x = player;
                break;
            case "c":
                state.c = player;
                break;
            default: throw new GameBoardException("Invalid key: " + key);
        }
        if (player == 0) state.lastPlayer = -1 * this.player;
        else state.lastPlayer = player;
        return state;
    }

    @Override
    protected State innerApply(State state) {
        return doIt(key, player, (State) state.clone());
    }

    @Override
    protected State innerUnapply(State state) {
        return doIt(key, 0, (State) state.clone());
    }

    @Override
    public String description() {
        return null;
    }

    @Override
    public String serialize() {
        return null;
    }

    @Override
    public Object deserialize(String s) {
        return null;
    }
}

File: src/main/scala/com/bytebreakstudios/tictactoe/scala/action/PlayAction.scala
package com.bytebreakstudios.tictactoe.scala.action

import com.bytebreakstudios.board.Action
import com.bytebreakstudios.board.utils.GameBoardException
import com.bytebreakstudios.tictactoe.scala.State

class PlayAction(val key:String, val player:Int) extends Action[State] {
  if (key == null) throw new GameBoardException("Key cannot be null")
  if (player == 0) throw new GameBoardException("Player must either be < 0 or > 0 but not == 0")
  if (!PlayAction.validKeys.contains(key.toLowerCase)) throw new GameBoardException(s"Invalid key: $key")


  override def innerApply(t: State): State = doIt(key.toLowerCase, Math.min(1, Math.max(-1, player)), t)

  override def innerUnapply(t: State): State = doIt(key.toLowerCase, 0, t)

  private def doIt(key:String, player:Int, state:State): State = {
    val tmp:State = key match {
      case "q" => state.copy(q = player)
      case "w" => state.copy(w = player)
      case "e" => state.copy(e = player)
      case "a" => state.copy(a = player)
      case "s" => state.copy(s = player)
      case "d" => state.copy(d = player)
      case "z" => state.copy(z = player)
      case "x" => state.copy(x = player)
      case "c" => state.copy(c = player)
      case _ => throw new GameBoardException(s"Invalid key: $key")
    }
    tmp.copy(lastPlayer = player match {
      case 0 => -1 * this.player
      case _ => player
    })
  }

  override def description(): String = ???

  override def serialize(): String = ???

  override def deserialize(s: String) = ???
}

object PlayAction{
  val validKeys = Seq("q", "w", "e", "a", "s", "d", "z", "x", "c")
}
